### PR TITLE
feat: add env driver

### LINKED ISSUE #23: add "env" driver that connects to $BUILDKIT_HOST
This driver allows creating a buildx instance directly pointing to the existing buildkit endpoint.

`docker buildx create --driver=env unix:///var/lib/buildkitd.sock`
`docker buildx create --driver=env` uses `BUILDKIT_HOST` env value with UNIX sock default.

I guess TLS info is just passed with a custom driver-opt:

```
docker buildx create --driver=env --driver-opt ca=mycafile.pem https://foobar
```

`rm/stop` commands are no-op in this driver.

A better name could be "remote".

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

Fixes #23 

cc: @AkihiroSuda @tonistiigi

### CODE REVIEW

File: commands/util.go [hunk 1]
```diff
@@ -117,7 +118,13 @@ func driversForNodeGroup(ctx context.Context, dockerCli command.Cli, ng *store.N
...
 				}
 
-				d, err := driver.GetDriver(ctx, "buildx_buildkit_"+n.Name, f, dockerapi, imageopt.Auth, kcc, n.Flags, n.Files, n.DriverOpts, n.Platforms, contextPathHash)
+				buildkitAPI, err := buildkitclient.New(ctx, endpoint)
+				if err != nil {
+					di.Err = err
+					return nil
+				}
+
+				d, err := driver.GetDriver(ctx, "buildx_buildkit_"+n.Name, f, dockerapi, buildkitAPI, imageopt.Auth, kcc, n.Flags, n.Files, n.DriverOpts, n.Platforms, contextPathHash, endpoint)
```
Comment (errordeveloper) [inline commands/util.go:127]: I think this functions needs some refactoring, the list of args is becoming very hard to read...

File: driver/manager.go [hunk 2]
```diff
@@ -51,6 +52,8 @@ type InitConfig struct {
 	// This object needs updates to be generic for different drivers
 	Name             string
 	DockerAPI        dockerclient.APIClient
+	BuildkitAPI      *buildkitclient.Client
+	BuldkitdAddr     string
```
Comment (errordeveloper) [inline driver/manager.go:56]: Is there a way to get the address from the client?

File: commands/create.go [hunk 3]
```diff
@@ -137,7 +137,8 @@ func runCreate(dockerCli command.Cli, in createOptions, args []string) error {
 		if len(args) > 0 {
 			ep, err = validateEndpoint(dockerCli, args[0])
 			if err != nil {
-				return err
+				logrus.Warnf("No valid Docker endpoint found, will continue with buildkitd endpoint.")
```
Comment (errordeveloper) [inline commands/create.go:140]: I'm not sure about this one, but I'm not yet familiar with the argument handling logic, perhaps something to look into.

File: commands/util.go [hunk 4]
```diff
@@ -231,7 +238,7 @@ func getInstanceByName(ctx context.Context, dockerCli command.Cli, instance, con
 	if err != nil {
 		return nil, err
 	}
-	return driversForNodeGroup(ctx, dockerCli, ng, contextPathHash)
+	return driversForNodeGroup(ctx, dockerCli, ng, contextPathHash, ng.Nodes[0].Endpoint)
```
Comment (errordeveloper) [inline commands/util.go:241]: This probably needs extra care, does `driversForNodeGroup` really need to take a concrete endpoint?

File: driver/manager.go [hunk 5]
```diff
@@ -51,6 +52,8 @@ type InitConfig struct {
 	// This object needs updates to be generic for different drivers
 	Name             string
 	DockerAPI        dockerclient.APIClient
+	BuildkitAPI      *buildkitclient.Client
```
Comment (tonistiigi) [inline driver/manager.go:55]: This is very driver-specific and shouldn't be here. Only the remote driver itself should return the client. For the address, not exactly sure why it is here. Maybe we need a generic "endpoint" field in here that is buildkit address for remote driver and docker address for container one.

File: commands/create.go [hunk 6]
```diff
@@ -137,7 +137,8 @@ func runCreate(dockerCli command.Cli, in createOptions, args []string) error {
 		if len(args) > 0 {
 			ep, err = validateEndpoint(dockerCli, args[0])
 			if err != nil {
```
Comment (tonistiigi) [inline commands/create.go:139]: If this is different things now for different drivers we should just move the validation to the driver level.

File: driver/remote/driver.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+func (d *Driver) Features() map[driver.Feature]bool {
+	return map[driver.Feature]bool{}
+}
+
+func (d *Driver) Factory() driver.Factory {
+	return d.factory
+}
+
+func (d *Driver) IsMobyDriver() bool {
+	return true
```
Comment (tonistiigi) [inline driver/remote/driver.go:56]: `false`

File: commands/util.go [hunk 8]
```diff
@@ -80,6 +81,13 @@ func driversForNodeGroup(ctx context.Context, dockerCli command.Cli, ng *store.N
 				defer func() {
 					dis[i] = di
 				}()
+
+				buildkitAPI, err := buildkitclient.New(ctx, n.Endpoint)
```
Comment (Dentrax) [inline commands/util.go:85]: i think we're initializing both `buildkitAPI` and `dockerapi` here if we set `--driver=env`. So `dockerapi` initializing unnecessarily. But couldn't find a better way.

**Comment by developer-guy** [comment]
cc: @crazy-max

File: driver/env/driver.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+func (d *Driver) Info(ctx context.Context) (*driver.Info, error) {
+	if d.BuldkitdAddr == "" && d.Driver == "env" {
+		return nil, errors.Errorf("buldkitd addr must not be empty")
+	}
+
+	c, err := client.New(ctx, d.BuldkitdAddr)
+	if err != nil {
+		return nil, errors.Wrapf(driver.ErrNotConnecting, err.Error())
```
Comment (AkihiroSuda) [inline driver/env/driver.go:32]: `fmt.Errorf` with `%w`

File: driver/env/driver.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+func (d *Driver) Rm(ctx context.Context, force, rmVolume, rmDaemon bool) error {
+	return fmt.Errorf("rm command is not implemented for this driver")
+}
+
+func (d *Driver) Client(ctx context.Context) (*client.Client, error) {
+	return client.New(ctx, d.BuldkitdAddr, client.WithSessionDialer(d.BuildkitAPI.Dialer()))
+}
+
+func (d *Driver) Features() map[driver.Feature]bool {
+	return map[driver.Feature]bool{}
```
Comment (AkihiroSuda) [inline driver/env/driver.go:57]: Why empty?

File: driver/env/driver.go (hunk 9 above)
Comment (tonistiigi) [reply driver/env/driver.go:32]: `fmt.Errorf` is not allowed because it doesn't store stacktrace

File: commands/util.go (hunk 8 above)
Comment (tonistiigi) [reply commands/util.go:86]: This is definitely not allowed. Only the `driver.Client()` is allowed to call this. `dockerapi` initialization is for the features like `--load` that move things to Docker.

 Another difference is that `clientForEndpoint` is just a struct setup, while `buildkitclient.New()` actually dials out to the endpoint.

**Comment by jedevc** [comment]
@developer-guy have you seen #1078? I've tried to finish off the work you started, do let me know what you think :smile:

**Comment by tonistiigi** [comment]
Merged in #1078

**Pull request closed without merging**